### PR TITLE
MESOS make launchGracePeriod configurable

### LINKED ISSUE #492: executor times out waiting for docker pull of large images
there's a hard-coded 5m timeout (`launchGracePeriod = 5 * time.Minute`) in executor.go and this is breaking the experience for some users that have huge docker images that can take a while to pull. this value should be configurable.

### PR DESCRIPTION
**Pull request by s-urbaniak**
This PR fixes https://github.com/mesosphere/kubernetes-mesos/issues/492

### CODE REVIEW

**Comment by s-urbaniak** [comment]
@sttts @jdef PTAL

File: contrib/mesos/pkg/executor/executor.go [hunk 1]
```diff
@@ -47,7 +47,6 @@ import (
 
 const (
 	containerPollTime = 1 * time.Second
```
Comment (s-urbaniak) [inline contrib/mesos/pkg/executor/executor.go]: Does it make sense to make this configurable too?

File: contrib/mesos/pkg/executor/executor.go [hunk 2]
```diff
@@ -47,7 +47,6 @@ import (
...
-	launchGracePeriod = 5 * time.Minute
 	podRelistPeriod   = 5 * time.Minute
```
Comment (s-urbaniak) [inline contrib/mesos/pkg/executor/executor.go]: Does it make sense to make this configurable too?

File: contrib/mesos/pkg/executor/service/service.go [hunk 3]
```diff
@@ -87,6 +88,7 @@ func (s *KubeletExecutorServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.SuicideTimeout, "suicide-timeout", s.SuicideTimeout, "Self-terminate after this period of inactivity. Zero disables suicide watch.")
 	fs.IntVar(&s.ShutdownFD, "shutdown-fd", s.ShutdownFD, "File descriptor used to signal shutdown to external watchers, requires shutdown-fifo flag")
 	fs.StringVar(&s.ShutdownFIFO, "shutdown-fifo", s.ShutdownFIFO, "FIFO used to signal shutdown to external watchers, requires shutdown-fd flag")
+	fs.DurationVar(&s.LaunchGracePeriod, "launch-grace-period", s.LaunchGracePeriod, "Launch grace period after which launching tasks will be cancelled . Zero disables launch cancellation.")
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: Is there such a flag in the kubelet as well?

Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: Then we could reuse it.

File: contrib/mesos/pkg/executor/config/config.go [hunk 4]
```diff
@@ -22,8 +22,9 @@ import (
...
 const (
-	DefaultInfoID         = "k8sm-executor"
-	DefaultInfoSource     = "kubernetes"
-	DefaultInfoName       = "Kubelet-Executor"
-	DefaultSuicideTimeout = 20 * time.Minute
+	DefaultInfoID            = "k8sm-executor"
+	DefaultInfoSource        = "kubernetes"
+	DefaultInfoName          = "Kubelet-Executor"
+	DefaultSuicideTimeout    = 20 * time.Minute
+	DefaultLaunchGracePeriod = 5 * time.Minute
```
Comment (jdef) [inline contrib/mesos/pkg/executor/config/config.go]: this probably needs to be referenced when the default config is being created

File: contrib/mesos/pkg/executor/service/service.go (hunk 3 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: Don't forget to add new flags to hack/verify-flags/known-flags.txt. 

As you haven't seen an error during committing, you probably don't have the kubernetes post-commit hook installed locally. You should. Catches a lot of errors.

**Comment by sttts** [comment]
There should be a --executor-launch-grace-period flag in the scheduler as well, passed down to the kubelet.

File: contrib/mesos/pkg/executor/config/config.go (hunk 4 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/config/config.go]: clarification: when flag defaults are assigned. See how DefaultSuicideTimeout is used (NewKubeletExecutorServer)

File: contrib/mesos/pkg/executor/executor.go (hunk 1 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: i wouldn't bother, this is probably going to be factored out at some point when we re-implement container monitoring

File: contrib/mesos/pkg/executor/executor.go (hunk 2 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: i don't think podRelistPeriod is in master anymore. rebase?

**Comment by s-urbaniak** [comment]
thanks @sttts @jdef for the pointers, will look into it

File: contrib/mesos/pkg/executor/executor.go (hunk 2 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: We still have the podStore in the executor, well hidden in the framework informer. So the constant still exists.

**Comment by erictune** [comment]
From shippable:
```
Found flags in golang files not in the list of known flags. Please add these to hack/verify-flags/known-flags.txt
launch-grace-period
```

File: contrib/mesos/pkg/executor/config/config.go (hunk 4 above)
Comment (s-urbaniak) [inline contrib/mesos/pkg/executor/config/config.go]: good catch, thanks

**Comment by s-urbaniak** [comment]
@sttts sorry, I don't find any pointers on timeouts in the kubelet. I'll test now locally with a very small timeout to see how this PR behaves.

**Comment by s-urbaniak** [comment]
@sttts @jdef PTAL.

I tested locally with a 100ms timeout and launched Pods end up (correctly?!) in a LOST state. The executor log grows indefinitely but I suggest to open a separate issue for that.

**Comment by sttts** [comment]
Can you create a ticket that the pods stay in "ImageNotReady" state? They should have "LaunchTimeout" or something similar as status reason.

**Comment by jdef** [comment]
What's wrong with ImageNotReady if the pull hasn't finished yet. Am I
missing something?

**Comment by sttts** [comment]
The pod has long died in the kubelet. This should be reflected. The reason is a free text field we can fill. I tthink we should make the difference between "still pulling" and "lost because pull took too long" clear in the reason.

**Comment by jdef** [comment]
this PR looks pretty good as-is. though i'm not crazy about adding two flags to the "known flags" txt file. would rather see a single flag 'mesos-launch-grace-period' that's shared across apps (kills two birds: namespaces the flag, and we only make 1 additional entry to known flags). in general our flags need an overhaul, but i'd rather see us not make the problem worse.

thoughts?

**Comment by sttts** [comment]
Those multiple flags really become a mess. We should clean that up with the generic multi sub-command flag handler. But not in this PR of course. `mesos-launch-grace-period` is fine with me.

**Comment by s-urbaniak** [comment]
@jdef @sttts PTAL

**Comment by jdef** [comment]
lgtm

**Comment by s-urbaniak** [comment]
@jdef squashed & mechanically rebased

**Comment by sttts** [comment]
Shippable failed.

**Comment by s-urbaniak** [comment]
flakes are hunting me ...

**Pull request merged**